File: 7/sol.js
const fs = require("fs");
const input = process.argv[2];
const lines = fs
  .readFileSync(input, "utf-8")
  .split("\n")
  .filter((x) => x !== "$ ls")
  .slice(1);

const root = {
  name: "",
  contents: [],
};
let curr = root;

for (let line of lines) {
  const dirMatch = line.match(/dir (.+)/);
  if (dirMatch !== null) {
    if (curr.contents.filter((x) => x.name !== dirMatch[1])) {
      curr.contents.push({
        name: dirMatch[1],
        contents: [],
        parent: curr,
      });
    }
    continue;
  }

  const fileMatch = line.match(/(\d+) (.+)/);
  if (fileMatch !== null) {
    curr.contents.push({
      name: fileMatch[2],
      size: parseInt(fileMatch[1]),
    });
    continue;
  }

  const cdMatch = line.match(/\$ cd (.+)/);
  if (cdMatch !== null) {
    if (cdMatch[1] == "..") {
      curr = curr.parent;
    } else {
      curr = curr.contents.filter((x) => x.name == cdMatch[1])[0];
    }
  }
}

const size = (node) => {
  if (node["contents"] === undefined) {
    return node.size;
  }

  return node.contents.reduce((total, node) => {
    return total + size(node);
  }, 0);
};

const dirsSmallerThan = (node, max) => {
  if (node["contents"] === undefined) {
    return [];
  }

  return node.contents
    .filter((x) => x.contents !== undefined)
    .filter((x) => size(x) <= max)
    .concat(node.contents.map((x) => dirsSmallerThan(x, max)).flat());
};

const part1 = dirsSmallerThan(root, 100000).reduce((total, node) => {
  return total + size(node);
}, 0);
console.log(`Sum of dirs less than 100,000: ${part1}`);

const spaceToClear = size(root) - 7e7 + 3e7;
const part2 = dirsSmallerThan(root, 100000000)
  .sort(function (a, b) {
    return size(a) > size(b) ? 1 : -1;
  })
  .filter((x) => size(x) >= spaceToClear)
  .map((x) => size(x))[0];
console.log(`Smallest dir for deletion: ${part2}`);

File: 7/sol.py
from __future__ import annotations
from enum import Enum
import math
import re
import functools

class File:
    def __init__(self, name: str, size: int):
        self.name = name
        self.size = int(size)

class Folder:
    def __init__(self, name: str, parent: Folder):
        self.name = name
        self.parent = parent
        self.contents = list()

class FileSystem:
    def __init__(self, input: str):
      lines = input.split('\n')
      lines = list(filter(lambda s: s != "$ ls", lines))
      lines.pop(0)      

      self.root = Folder("", None)
      curr = self.root

      for line in lines:
        dirMatch = re.search(r"dir (.+)", line)
        if dirMatch is not None:
          curr.contents.append(Folder(dirMatch.group(1), curr));
          continue

        fileMatch = re.search(r"(\d+) (.+)", line)
        if fileMatch is not None:
          curr.contents.append(File(fileMatch.group(2), fileMatch.group(1)));
          continue

        cdMatch = re.search(r"\$ cd (.+)", line)
        if cdMatch is not None:
          if cdMatch.group(1) == "..":
            curr = curr.parent
          else:
            curr = list(filter(lambda s: s.name == cdMatch.group(1), curr.contents))[0]
          continue

def size(node):
  if isinstance(node, File):
    return node.size
  return functools.reduce(lambda total, node: total + size(node), node.contents, 0)

def isFolder(node):
  return isinstance(node, Folder)

def dirsSmallerThan(node, max):
  if isinstance(node, File):
    return []

  return functools.reduce(
    lambda acc, node: acc + ([node] if size(node) < max else []) + dirsSmallerThan(node, max),
    list(filter(isFolder, node.contents)),
    []
  )

if __name__ == '__main__':
    with open('input.txt') as f:
      fileSystem = FileSystem(f.read())
      
      part1 = functools.reduce(
        lambda acc, folder: acc + size(folder),
        dirsSmallerThan(fileSystem.root, 1e5),
        0
      )
      print(f"Sum of dirs less than 100,000: {part1}")

      spaceToClear = size(fileSystem.root) - 7e7 + 3e7;
      part2 = size(list(filter(
        lambda x: size(x) >= spaceToClear,
        sorted(dirsSmallerThan(fileSystem.root, 100000000), key=lambda x: size(x))
      ))[0])
      print(f"Smallest dir for deletion: {part2}")